### ISSUE TITLE
How to handle type-awareness disparity between DFD and `VkFormat`

### ISSUE THREAD

**Issue by MarkCallow** [KhronosGroup/KTX-Specification#14]
The Data Format Descriptor has no awareness of primitive types. A DFD describes the data as it is laid out in memory and is thus independent of endianness. However Vulkan & GL formats are type aware and, when the file writer and reader have different endianness, those with primitive type sizes of 2, 4 or 8 bytes must be byte-swapped into the reader's endianness before the data can be uploaded to GL or Vulkan using the vkFormat specified in the KTX file.

Note that there are a few VK_FORMATS which have opposite-endian equivalents that would enable uploads to be done without any byte swapping, e.g `VK_FORMAT_R8G8B8A8_UNORM` and `VK_FORMAT_A8B8G8R8_UNORM_PACK32`. These cases are very much the minority so it doesn't change the need to resolve this issue.

Changing the endianness of DFD-described data requires

- recognizing that the data has a size that requires swapping which is somewhat involved.
- modifying the DFD to describe the changed-endian data
- byte-swapping the data.

Andrew G. has promised some sample code to do this.

The DFD spec. is supporting a lot wider range of data layouts than KTX2 needs which means Andrew finds making DFD type aware unattractive.  However placing restrictions on the DFDs used in KTX files is theoretically possible. I'm not sure what restrictions we could make in practice that would help here.

Meanwhile the KTX2 header is also type-aware. It has 32- and 64-bit fields so we can't remove the `endianness` field and swapping requirement we already have.

How to we handle the disparity?

1. Require that the image data should always be little-endian. Drawbacks here are that big-endian systems would have to swap bytes on both writing and reading KTX files and writers on big-endian machines are very likely to overlook this requirement leading to invalid files.
2. When the endianness field indicates a mismatch require readers to determine from the DFD if the data needs swapping, byte swap it and rewrite the DFD.
3. Retain a `typesize` field which is used when `vkFormat != VK_UNKNOWN_FORMAT`.  In this case the DFD must be rewritten as well. When it is VK_UNKNOWN_FORMAT` let the reader decide how it wants to handle the data.
4. Some other way?

Subject to seeing the code sample from Andrew, I am leaning towards 2.

**Comment by jherico** [comment]
I'm a little confused...

> Retain a typesize field which is used when vkFormat != VK_UNKNOWN_FORMAT.

is that supposed to be `==` rather than `!=`?  Why would you need a typesize as well as well defined vkFormat, if vkFormat always implies a specific type size?

I'm kind of tempted by option 1.  All the other options involve having to code for the potential for doing byte swapping work on all platforms, because you never know if there will be a mismatch.  Option 1 means that you don't have to deal with potential byte swapping, but will instead _know_ whether you'll need to byte swap at compile time.

**Comment by fluppeteer** [comment]
> Andrew G. has promised some sample code to do this.

Privately I promised to _try_! But I'll do my best. Detecting when adjacent fields can be merged after a byte swap isn't especially difficult; detecting whether byte swapping would lead to two things being identical is marginally harder. There's arguably a case for folding generic byte swizzle behaviour into the same function...

There's a distinction between endianness of the peripheral information (including the data format descriptor) and that of the underlying image. I think I suggested that the data format descriptor itself should be little-endian when a portable representation is required barring other constraints (which KTX2 does have), but that endianness should otherwise match the system to maximise processing speed; that language is due to be cleaned up, not least because some fields need to move a bit in an updated revision. So I think having a native-endian wrapper (with 32-bit and 64-bit fields given an explicit endianness) doesn't necessarily mandate anything about the actual image content, or vice-versa.

I'm inclined to say that:

- Unless a system only ever expects to see data which it has written itself, a big-endian system is likely to need to handle little-endian content - meaning that it will already need endianness swapping code (at least on read).
- Most modern systems (notably arm and Intel) are usually little-endian, so mandating a little-endian representation might simplify the majority of implementations.
- A typesize field (at least for common formats) would simplify conversions considerably, since (without a list of "supported formats") there is often more than one possible interpretation of the same data. In this case, a big-endian reader can either look at the Vulkan format and say "oops, I need to swap that" or it can look at the data format descriptor and spot a format that crosses a byte boundary; in some cases it will need explicit rearrangement code, and in others it might simply choose a different format which happens to give the same bytes the same meaning. If we also allow texture channel swizzles, this approach becomes more flexible.

In other words, I'd like 1 + 3 (= 4!) where the typesize field is always written, but can be ignored by little-endian formats. Of course, if you have the Vulkan types, you can deduce the type sizes anyway from a look-up table, there's just no programmatic way to deduce them.

If the data needs to be reordered in memory in comparison with the file in order to be accessed efficiently on a big-endian system (=2), if the DFD is still expected to be interpreted by the user as a description of the memory, that DFD will need to be updated. That's likely to be a simple chop-the-channel operation -  if we started in a little-endian representation, it's highly unlikely that channels would have to be merged again.

So I guess I favour "all of the above".

**Comment by MarkCallow** [comment]
> is that supposed to be == rather than !=? Why would you need a typesize as well as well defined vkFormat, if vkFormat always implies a specific type size?

It's to avoid that code snippet we've been discussing over in issue #8. I am concerned how to document the correct typesize. Either a code snippet or a table of formats & type sizes will require constant updates when Vulkan Core or extensions add new formats. An explicit `typesize` field eliminates those issues. I suppose we could put neither in the spec. and instead say if the format has a primitive type of 2, 4 or 8 bytes as indicated by its name, that byte swapping must be done.

**Comment by MarkCallow** [comment]
> I think I suggested that the data format descriptor itself should be little-endian when a portable representation

Yes. That is my no. 1.

**Comment by fluppeteer** [comment]
> I suppose we could put neither in the spec. and instead say if the format has a primitive type of 2, 4 or 8 bytes as indicated by its name, that byte swapping must be done.

That works for me (so long as we continue to ensure the Vulkan formats are appropriately named - and I believe explicit sizing information is coming in the Vulkan spec). It does mean that someone dealing only with OpenGL needs to pay a bit more attention to Vulkan, unfortunately. That said, it may be possible to do a GL/Vulkan mapping descriptively without too much effort.

> Yes. That is my no. 1.

&lt;Monty Python&gt;No it isn't.&lt;/Monty Python&gt;

Your no.1 was that the image data (at least, I assume this means "memory representing each pixel") has to be encoded little-endian. I was arguing that the layout of the texels is independent of the layout of the metadata - you could say an image is 1024x1024 pixels with the value "1024" represented in big-endian form (aka 0x04, 0x00) while still representing every pixel in the image with a little-endian encoding.

I was only making the case that having 32-bit and 64-bit fields in the KTX2 header didn't necessarily mandate the ability to have multiple endianness for the texel data - I _do_ think it's a good idea to mandate that the texel data itself is little-endian, but I didn't want to conflate arguments.

**Comment by MarkCallow** [comment]
@fluppeteer that's what I was trying to get at in no. 1. The KTX header data can have differenet endianness than the texel data. I think we're in agreement.

**Comment by fluppeteer** [comment]
@MarkCallow ah, sorry, I thought you were indicating that having endianness for the header dictated (or encouraged) endianness support for the main data. I would equally advocate mandating that the header be stored in little-endian format, and let big-endian systems swap the fields (whose size should be known implicitly) if needed - partly to simplify the header, partly to simplify the implementation on more common systems, and partly because big-endian developers will already be familiar with having to do this for other formats!

Out of interest, how would you expect endianness to be handled for 128-bit formats such as ASTC or BC2? BC2 describes itself as 64 bits of alpha followed by 64 bits of colour, suggesting it's two 64-bit words which might need to be endian-swapped. ASTC seems to be raw 128-bit. I'm unsure how a big-endian 32-bit system would handle a 64-bit quantity, or how either would handle an 128-bit quantity. I guess if it's stored little-endian, it's the problem of the loader on a big-endian system to do whatever the architecture might consider to be "the right thing" - I guess I should ask someone what arm considers the bit layout of ASTC to be when their chips are configured in big-endian mode.

**Comment by fluppeteer** [comment]
FYI, I asked arm. ASTC content was not endian-swapped when the GPU was configured to be big-endian.

**Comment by lexaknyazev** [comment]
It seems that the case for big-endian texel data is getting smaller, right? Which GPUs do expect big-endian data?

**Comment by fluppeteer** [comment]
I imagine that an arm GPU configured to be big-endian (when this was an option) accessing an R16 value would still have byte swapped it accordingly - it's just that ASTC never got swapped. I'm seeking input on whether BCn ever got swapped (which I guess might have been relevant to MIPS Windows NT?) and ETC is explicit about its byte order, but the uncompressed multi-byte formats are still an issue.

I don't believe any company currently using big-endian hardware (at least by default - several systems are configurable-endian) currently has Khronos adopter status. Sadly I suspect this doesn't mean people won't try to wrap really old textures in KTX2...

**Comment by MarkCallow** [comment]
I am struggling to see the benefit of DFD's current mechanism whereby it describes the current layout in memory on a per-byte basis starting with the LSB of each byte. I agree that big-endian machines are in the minority but the current scheme makes in very difficult to do most (all?) image processing operations on a big-endian machine. The most striking examples are with the *packed* formats, e.g, RGB565. Here the descriptor will show the first 3 bits of G at the start (LSBs) of the first byte and the second 3 at the end (MSBs) of the second byte. It seems to me that a lot of understanding and a lot of bit twiddling is necessary before the application can extract, e.g, the G bits. If the application wants to process the image data using 16-bit types, both the DFD and the data need to be converted. (If the data's primitive type is 16-, 32-, or 64-bit applications will for sure want to conduct image processing operations using those types, especially the last 2.)

If instead the DFD described, for the above example,  a 16-bit quantity starting from the LSBs then I think it is much easier to understand. You've got 5 bits of B followed by 6 bits of G followed by 5 bits of R. Masking is simple to understand. The DFD remains the same regardless of endianness. Yes the data needs to be swapped when transferred between machines of opposite endianness but that is a straightforward operation.

@fluppeteer what do you see as the advantages of the current mechanism? Perhaps if I saw some sample code for calculating the primitive size of the data and for extracting specific bits it would allay my fears.

**Comment by MarkCallow** [comment]
@fluppeteer, one more question. If a DFD describes a 32-bit component, e.g, 4 "channels" of red, how does an app that wants to do arithmetic on the values determine the described byte order and whether it matches the local machine?

I could probably find the answer in the DFD spec. but I'm under the weather right now so would appreciate guidance from its author.

**Comment by dewilkinson** [comment]
Reviewed in 12/21 call

In the 12/20 3DFormats call, no objections to restricting support to LE only were raised. Additionally it was noted that glTF already requires  binary data to be LE.

Given no objections by any of the group or external stakeholders, KTX2 will assume LE format for all data fields and binary chunks

**Comment by dewilkinson** [comment]
Ready to close

**Comment by MarkCallow** [comment]
I need to do the spec. work before closing.

**Comment by MarkCallow** [comment]
Fixed by PR #60.
